### ISSUE TITLE
[BUG] AppObserver Concurrency bug: Collection was modified; enumeration operation may not execute

### ISSUE THREAD

**Issue by Jaans** [microsoft/service-fabric-observer#184]
**Describe the bug**
FO is throwing exceptions when reporting health.
The below is a stack trace found in the log files at: `C:\observer_logs\ObserverManager`

```
2021-11-08 00:24:40.3433--WARN--Handled AggregateException from AppObserver:
System.AggregateException: One or more errors occurred. (One or more errors occurred. (Collection was modified; enumeration operation may not execute.))
 ---> System.AggregateException: One or more errors occurred. (Collection was modified; enumeration operation may not execute.)
 ---> System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   at System.Linq.Enumerable.Average(IEnumerable`1 source)
   at FabricObserver.Observers.Utilities.FabricResourceUsageData`1.get_AverageDataValue() in D:\a\1\s\FabricObserver.Extensibility\Utilities\FabricResourceUsageData.cs:line 138
   at FabricObserver.Observers.ObserverBase.ProcessResourceDataReportHealth[T](FabricResourceUsageData`1 data, T thresholdError, T thresholdWarning, TimeSpan healthReportTtl, HealthReportType healthReportType, ReplicaOrInstanceMonitoringInfo replicaOrInstance, Boolean dumpOnError) in D:\a\1\s\FabricObserver.Extensibility\ObserverBase.cs:line 1180
   at FabricObserver.Observers.AppObserver.<>c__DisplayClass43_0.<ReportAsync>b__0(ReplicaOrInstanceMonitoringInfo repOrInst, ParallelLoopState state) in D:\a\1\s\FabricObserver\Observers\AppObserver.cs:line 369
   at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
   at System.Threading.Tasks.TaskReplicator.Replica.Execute()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.TaskReplicator.Run[TState](ReplicatableUserAction`1 action, ParallelOptions options, Boolean stopOnFirstFailure)
   at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Parallel.ThrowSingleCancellationExceptionOrOtherException(ICollection exceptions, CancellationToken cancelToken, Exception otherException)
   at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.ForEachWorker[TSource,TLocal](IEnumerable`1 source, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Action`3 bodyWithStateAndIndex, Func`4 bodyWithStateAndLocal, Func`5 bodyWithEverything, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.ForEach[TSource](IEnumerable`1 source, ParallelOptions parallelOptions, Action`2 body)
   at FabricObserver.Observers.AppObserver.ReportAsync(CancellationToken token) in D:\a\1\s\FabricObserver\Observers\AppObserver.cs:line 142
   at FabricObserver.Observers.AppObserver.ObserveAsync(CancellationToken token) in D:\a\1\s\FabricObserver\Observers\AppObserver.cs:line 133
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait(TimeSpan timeout)
   at FabricObserver.Observers.ObserverManager.RunObserversAsync()
```

**To Reproduce**
Deploy the latest Service Fabric Observer using default config to a fresh 1-node 'OneBox' Development Cluster using the latest version of Service Fabric v8.2.1235.9590)

**Expected behavior**
Exceptions to not be thrown.

**Desktop (please complete the following information):**
 - OS: Windows 10 PRO
 - Version: 21H1 (OS Build 19043.1288)

**Additional context**

**Comment by GitTorre** [comment]
Thanks for reporting this. Will look into it and fix.

**Comment by GitTorre** [comment]
I installed the latest FO (3.1.20) on a one node cluster (dev machine) with a fresh install of SF Runtime 8.2.1235.9590. I am unable to repro this. The OS is Windows 11, but I don't think the OS is the problem. That exception means that a collection was changed while an iteration over it was taking place (typically, this would be a foreach loop). Looking at the stack, my hypothesis is that the collection (Data property of FRUD) was changed while the Average was being computed. 

Can you consistently reproduce this behavior?

**Comment by GitTorre** [comment]
I have been unable to reproduce this all day. If you have a consistent repro, perhaps we can get on a Teams call and you can show me, we can debug on your machine.

**Comment by Jaans** [comment]
Thanks for the follow-up. I too would doubt that the issue lies with Windows 11.

This is still ongoing:
![image](https://user-images.githubusercontent.com/3510981/140839504-d24fe3c9-dc49-45f4-bacb-89dcc6165cdd.png)

We grabbed `Microsoft.ServiceFabricApps.FabricObserver.Windows.FrameworkDependent.3.1.20.sfpkg` from the GitHub releases and deployed via PowerShell command line.

A difference that might be relevant is that we configure our SF Cluster with the following values:
* Hosting.EnableActivateNoWindow = true
* ReconfigurationAgent.SendLoadReportInterval = 60

![image](https://user-images.githubusercontent.com/3510981/140840250-704f8749-a3c6-4601-9ee1-2b9a53f5d270.png)

![image](https://user-images.githubusercontent.com/3510981/140840296-c4325f29-b601-402f-a65b-e480d77b42d5.png)

Let me know if I can help further. Perhaps a memory dump?

Thanks

**Comment by GitTorre** [comment]
Can you zip up your log files, including AppObserver.log and send to me or provide with a link to the zip file on a cloud drive? 

ctorre at microsoft.com 

Thanks

**Comment by GitTorre** [comment]
I updated node configuration with the same two settings you mention above (I don't see how these are related to this issue). I also deployed FO 3.1.20 FrameworkDependent from the SFPKG on the repo. Still no repro. 

What version of .NET Core is installed on your one-node cluster? Since you are using FrameworkDependent, this means you know that the same version of .NET Core is installed on the target machine. 

Can you try with SelfContained and see if this repros?

**Comment by Jaans** [comment]
Sure, I've emailed you a ZIP archive of the logs.

With “framework dependent” we are running .NET Core 3.1.414.
Will try “self contained” version and let you know the outcome of that later this evening.

Thank you

**Comment by GitTorre** [comment]
Thanks for all the logs. You are not doing anything wrong nor do you have any misconfigurations or incorrect runtime version...

This is a concurrency bug. 

Until I fix it, please just disable AppObserverEnableConcurrentMonitoring like this:
```xml
    <!-- Concurrency/Parallelism Support -->
    <Parameter Name="AppObserverEnableConcurrentMonitoring" DefaultValue="false" />
```

and change MonitorDuration to 1 like this:
```xml
    <Parameter Name="AppObserverMonitorDuration" DefaultValue="00:00:01" />
```
in ApplicationManifest.xml. 

**Thank you for reporting this**. Since the mitigation is to simply disable concurrent monitoring in AppObserver, you should be set to move forward. Please do remember to keep the MonitorDuration at 1 as it applies to each service process. You can check back on this Issue for updates.

**Comment by Jaans** [comment]
Awesome, that works. 
Thank you for looking into this.

**Comment by GitTorre** [comment]
Update: There is a fix for this transient concurrency issue in the [develop](https://github.com/microsoft/service-fabric-observer/tree/develop) branch. This will ship in 3.1.21. Please clone this repo, change branch to develop, build, deploy to your test machine and see if this happens again. Thank you!

**Comment by GitTorre** [comment]
Update 3: FO 3.1.21 RC4 is now in develop branch. More performance improvements.

**Comment by GitTorre** [comment]
Fixed in 3.1.21
